internal/master/service.go

```diff
@@ -511,10 +563,58 @@ func (ds *distributedService) GetStatus(context context.Context, request *emptyp
 			regionInfo.Status = serverpb.RegionStatus_PRIMARY_FOLLOWER
 		}
 	}
-	ds.lg.Debug("Current Info", zap.String("Status", regionInfo.Status.String()))
+	ds.opts.Logger.Debug("Current Info", zap.String("Status", regionInfo.Status.String()))
 	return regionInfo, nil
 }
 
+func (ds *distributedService) Check(ctx context.Context, healthCheckReq *serverpb.HealthCheckRequest) (*serverpb.HealthCheckResponse, error) {
+	regionInfo := ds.DKVService.(*standaloneService).regionInfo
+	if ds.isClosed {
+		return &serverpb.HealthCheckResponse{Status: serverpb.HealthCheckResponse_NOT_SERVING}, nil
+	} else {
+		leaderId, members := ds.raftRepl.ListMembers()
+		selfId := ds.raftRepl.Id()
+		isFollower := members[selfId] != nil && (members[selfId].Status == models.NodeInfo_FOLLOWER)
+		if leaderId == selfId || isFollower {
+			return &serverpb.HealthCheckResponse{Status: serverpb.HealthCheckResponse_SERVING}, nil
+		} else if !isFollower {
+			return &serverpb.HealthCheckResponse{Status: serverpb.HealthCheckResponse_NOT_SERVING}, nil
+		}
+	}
+	ds.opts.Logger.Debug("Current Info", zap.String("Status", regionInfo.Status.String()))
+	return &serverpb.HealthCheckResponse{Status: serverpb.HealthCheckResponse_NOT_SERVING}, nil
+}
+
+func (ds *distributedService) Watch(req *serverpb.HealthCheckRequest, watcher serverpb.HealthCheck_WatchServer) error {
+	if ds.isClosed {
+		res, err := ds.Check(context.Background(), req)
+		if err != nil {
+			return err
+		}
+		return watcher.Send(res)
+	}
+	ticker := time.NewTicker(time.Duration(ds.opts.HealthCheckTickerInterval) * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			res, err := ds.Check(context.Background(), req)
+			if err != nil {
+				return err
+			}
+			if err := watcher.Send(res); err != nil {
+				return err
+			}
+		case <-ds.shutdown:
+			res, err := ds.Check(context.Background(), req)
+			if err != nil {
+				return err
+			}
```

Comment:
Should we not send a shutdown/NOT_SERVING instead?

Reply:
If the `ds` instance is available (which should be in this case), the next line will take care of sending NOT_SERVING response.